File: ios-native/Jottery/Jottery/Database/EncryptionRepository.swift
import Foundation
import GRDB

/// Reads and writes encryption metadata (salt, iterations, algorithm).
struct EncryptionRepository: Sendable {

    let db: DatabaseManager

    /// Fetch the encryption metadata (returns nil if vault not set up).
    func get() throws -> EncryptionMetadata? {
        try db.dbPool.read { db in
            try EncryptionMetadata.fetchOne(db, key: 1)
        }
    }

    /// Store encryption metadata for a new vault.
    func store(_ metadata: EncryptionMetadata) throws {
        try db.dbPool.write { db in
            try metadata.save(db)
        }
    }

    /// Check if a vault has been set up.
    func isVaultSetUp() throws -> Bool {
        try get() != nil
    }

    /// Save envelope encryption metadata, clearing legacy salt/iterations.
    func saveEnvelope(envelopeVersion: Int, deviceSalt: String,
                      localWrappedMaster: String, wrappingKdfVersion: Int) throws {
        try db.dbPool.write { db in
            try db.execute(sql: """
                UPDATE encryption_metadata
                SET envelope_version = ?, device_salt = ?,
                    local_wrapped_master = ?, wrapping_kdf_version = ?,
                    salt = '', iterations = 0
                WHERE id = 1
            """, arguments: [envelopeVersion, deviceSalt, localWrappedMaster, wrappingKdfVersion])
        }
    }
}

File: android-native/app/src/main/java/com/jottery/android/database/repository/EncryptionRepository.kt
package com.jottery.android.database.repository

import com.jottery.android.database.dao.EncryptionMetadataDao
import com.jottery.android.model.EncryptionMetadata

/**
 * Repository for encryption metadata (salt, iterations, verification token).
 */
class EncryptionRepository(
    private val dao: EncryptionMetadataDao,
) {
    suspend fun get(): EncryptionMetadata? = dao.get()

    suspend fun store(metadata: EncryptionMetadata) = dao.insertOrReplace(metadata)

    /** Save envelope encryption metadata, clearing legacy salt/iterations. */
    suspend fun saveEnvelope(
        envelopeVersion: Int,
        deviceSalt: String,
        localWrappedMaster: String,
        wrappingKdfVersion: Int,
    ) {
        val existing = get() ?: return
        val updated = existing.copy(
            salt = null,
            iterations = null,
            envelopeVersion = envelopeVersion,
            deviceSalt = deviceSalt,
            localWrappedMaster = localWrappedMaster,
            wrappingKdfVersion = wrappingKdfVersion,
        )
        dao.insertOrReplace(updated)
    }
}
